Add DebugBar interface

jespino opened the pull request:
#### Summary
This PR adds the debug bar interface

#### Ticket Link
Noticked

#### Related Pull Requests
- Has server changes mattermost/mattermost-server#22410

#### Screenshots

#### Release Note
```release-note
Added debugbar to improve the development experience
```

jespino committed:
Adding debugbar initial interface

jespino committed:
Improving debugbar view

jespino committed:
Adding filtering support

jespino committed:
Adding the debugbar/systeminfo endpoint

jespino committed:
Improving the debug bar architecture

jespino committed:
Reorganizing debugbar code

jespino committed:
Adding logs to the debugbar

jespino committed:
Using the DebugBar client config provided by the server

jespino committed:
Add the store calls parameters to the interface

jespino committed:
Merge remote-tracking branch 'origin/master' into debugbar

jespino committed:
Adding debugbar email traces

jespino committed:
Making the code more type aware and cleaner

jespino committed:
Improving types

jespino committed:
Improving a bit the layout of the information in the table

jespino committed:
Adding explain support

koox00 committed:
Updates debugbar styling

koox00 committed:
Fixes multiline Code

koox00 committed:
Styles email sent

jespino committed:
Adding SQL Queries modal with explain info

jespino committed:
Adding feature to resize the bar

jespino committed:
Adding virtualized lists

jespino committed:
Adding email details

jespino committed:
Changing the width of the method to work well with DELETE

jespino committed:
Fixing bug in SQL queries with arguments

jespino committed:
Adding the server log message

jespino committed:
Removing the small from the server log message

jespino committed:
Merge remote-tracking branch 'origin/master' into debugbar

jespino committed:
Fix linter errors

neallred commented on components/debugbar/emailssent.tsx:
```diff
@@ -0,0 +1,102 @@
+    const [viewEmail, setViewEmail] = useState<DebugBarEmailSent|null>(null);
+    let emails = useSelector(getEmailsSent);
+    if (filter !== '') {
+        emails = emails.filter((v) => JSON.stringify(v).indexOf(filter) !== -1);
```
This works for now. Wanted to point out that some of the fields of the email will be identical across emails like SMTP config, and it will include html content so searching things like html tag names will return matches.

*If* we wanted to refine in the future we could do something like parsing the html document and grabbing the inner text so that only the text portion is searchable.

```tsx
const el = document.createElement( 'html' );
el.innerHTML = v.htmlBody;
contentToSearch = `${v.category} ${v.cc} ${v.inReplyTo} ${v.messageID} ${v.references} ${v.subject} ${v.time} ${v.to} ${v.type} ${el.innerText}`

```

And then again, maybe we *want* to be able to search for html class names that appear (or fail to appear)

jespino replied:
For now is going to be like this, after this is merged, we can improve this using something like https://lunrjs.com/ or (elastic lunr)

neallred commented on components/debugbar/logs.tsx:
```diff
@@ -0,0 +1,74 @@
+                title={JSON.stringify(data[index].fields)}
+            >
+                <Code
+                    code={JSON.stringify(data[index].fields)}
```
Nit: There's enough server logs that it might be worth stringifying the data only once and storing it in a constant. That said, things feel reasonably snappy on my local at the moment.

jespino replied:
I think that we can use a client side indexing library and subscribe to the store to generate the indexes, that way we also have way better search support

neallred commented on components/debugbar/apicalls.tsx:
```diff
@@ -0,0 +1,80 @@
+                itemCount={calls.length}
+                itemSize={50}
+                height={height}
+                width={window.innerWidth - 2}
```
reading window.innerWidth is among the [things causing reflow](https://gist.github.com/paulirish/5d52fb081b3570c81e3a). If we start notice things being sluggish, it might be worth checking if calculating window.innerWidth at the top in a hook and passing down as a prop can speed things up. Something like https://usehooks-ts.com/react-hook/use-window-size . I'm especially thinking of frequent ongoing tabs like server logs and api calls.

jespino replied:
Fixed

neallred reviewed (changes requested):
Code looks good overall! Most of the comments are nits or suggested polish, mostly around theming, i18n, and possible perf optimizations. The ones I think we definitely should fix prior to merging are:
* add the types to the reducers
* fix the event listener leak

jespino commented:
Thanks @neallred really helpful comments, I'll implement the changes today, I'll skip the i18n changes for now, I'm not sure if we want to add load to the translators when it is not strictly needed

jespino commented on components/debugbar/debugbar.tsx:
```diff
@@ -0,0 +1,141 @@
+
+    if (hidden) {
+        return (
+            <button
```
This leads to another conversation, do we wand to add that translation load effort for all the strings inside the debugbar for something that is not user facing (only developers facing)? I would keep this untranslated for now, and add the translations in a later stage.

jespino commented on components/debugbar/debugbar.scss:
```diff
@@ -0,0 +1,187 @@
+@import 'utils/mixins';
+
+.DebugBarButton {
+    position: fixed;
```
ternary has a transparent background, so I used primary, that works fine there too :)

jespino committed:
Some styles improvements and making it themes aware

jespino commented on components/debugbar/debugbar.scss:
```diff
@@ -0,0 +1,187 @@
+    background: white;
+    grid-row: 4;
+    grid-column: 1/span 5;
+    border: 1px solid #aaa;
```
I did some work around making it more aware of the CSS variables, now it should fit better with the whole styling of the mattermost application, and also is theme aware :)

jespino committed:
improving a bit the look and feel

jespino commented on components/debugbar/debugbar.scss:
```diff
@@ -0,0 +1,187 @@
+
+    &__Tab {
+        + .DebugBar__Tab {
+            border-left: 1px solid rgba(0,0,0, 0.07);
```
I have entirely removed that but I'm doing something similar in other way

jespino commented on components/debugbar/debugbar.scss:
```diff
@@ -0,0 +1,191 @@
+    .DebugBarTable {
+        width: 100%;
+        background: var(--center-channel-bg);
+        font-family: "Times New Roman", Times, serif;
```
Removed the font and using the standard one

jespino committed:
Using classnames package in another place

jespino commented on components/debugbar/debugbar.tsx:
```diff
@@ -0,0 +1,162 @@
+            <div className='DebugBar__Tab'>
+                <button
+                    key={tab}
+                    className={'header__Button ' + (isSelected(tab) ? 'selected' : '')}
```
done

jespino commented on components/debugbar/debugbar.tsx:
```diff
@@ -0,0 +1,162 @@
+    {tab: 'sql', text: 'SQL Queries'},
+    {tab: 'logs', text: 'Server'},
+    {tab: 'emails', text: 'Emails'},
+    {tab: 'system', text: 'System Info'},
```
Same thing, I would defer that until we make a conscious decision about using translations for this area of the application. I don't think is needed and I think is going to add extra load to the translators.

jespino committed:
Improving a bit the code for the debugbar tabs

jespino commented on components/debugbar/debugbar.tsx:
```diff
@@ -0,0 +1,162 @@
+        return (
+            <div className='DebugBar__Tab'>
+                <button
+                    key={tab}
```
fixed by restructuring a bit the code.

jespino committed:
Fixing the leak in the event handler

jespino committed:
Adding a minimum size of 34 pixels to the debugbar

jespino commented on components/debugbar/debugbar.tsx:
```diff
@@ -0,0 +1,162 @@
+                className='handler'
+                draggable={true}
+                onDragEnd={() => {
+                    document.removeEventListener('dragover', (e) => {
```
done

jespino commented on components/debugbar/debugbar.tsx:
```diff
@@ -0,0 +1,162 @@
+                }}
+                onDragStart={() => {
+                    document.addEventListener('dragover', (e) => {
+                        setHeight(window.innerHeight - e.pageY);
```
done, good point

jespino committed:
Removed unused class

jespino commented on components/debugbar/debugbar.tsx:
```diff
@@ -0,0 +1,162 @@
+                    <QuickInput
+                        id='searchChannelsTextbox'
+                        placeholder='Filter'
+                        className='form-control filter-textbox'
```
done

jespino commented on components/debugbar/debugbar.tsx:
```diff
@@ -0,0 +1,162 @@
+                {tab !== 'system' && (
+                    <QuickInput
+                        id='searchChannelsTextbox'
+                        placeholder='Filter'
```
defered until we decide what we do with i18n

neallred reviewed (approved):
LGTM, thanks @jespino and @koox00 both for implementing! Looks awesome and I think this will give us a lot more insight a lot more easily into the impact our code changes have on the system.

Last thing I noticed, looks like the System Info tab doesn't yet support theming:
![system-info-theming](https://user-images.githubusercontent.com/13738432/222454391-19bf9b09-c642-4ed8-9000-3411cd5b8f4f.png)

jespino commented on components/debugbar/systeminfo.tsx:
```diff
@@ -0,0 +1,35 @@
+    const [systemInfo, setSystemInfo] = useState<any>(null);
+
+    useEffect(() => {
+        const interval = setInterval(() => {
```
Fixed

jespino committed:
Adding types to the debugbar reducers

jespino commented on packages/mattermost-redux/src/reducers/entities/debugbar.ts:
```diff
@@ -0,0 +1,96 @@
+import {DebugBarState} from '@mattermost/types/debugbar';
+import {DebugBarTypes} from 'mattermost-redux/action_types';
+
+export function emailsSent(state: any[] = [], action: GenericAction): any[] {
```
Good call, done

jespino committed:
Fixing system info theming support

jespino commented:
@neallred thanks, I already fixed the theming problem.

jespino committed:
Putting debugbar on top of everything

jespino commented on components/debugbar/debugbar.scss:
```diff
@@ -0,0 +1,191 @@
+
+.DebugBarButton {
+    position: fixed;
+    z-index: 1000;
```
I added the changes to put the debug bar and the debugbar button always on top

jespino committed:
Fix reflow with the window.innerHeight

jespino committed:
Fix reflow with the window.innerHeight

jespino committed:
Fixing linter errors

koox00 committed:
Collapse expand the bar, and full query modal

koox00 committed:
Styles the sql query modal

koox00 committed:
Minor refactor

koox00 committed:
Fixes styling

koox00 committed:
Satisfy linter

koox00 committed:
[Debugbar]: adds search in individual tabs (#12316)

* Adds search in individual tabs

Adds search in sql queries and server logs tabs.
Refactors components in a component folder.
Fixes some styling issues.
Adds an Empty state component.
Adds the ability to call clear for an individual tab.

* Adds regex filter to api and store calls

* Fixes linter

* Refactors selectors to filter as well

* Removes unneeded prop

* Fixes typo

* Highlights selected query in the list

* Adds info level filter server logs

jespino commented:
Migrated to the monorepo
